### PR TITLE
Add APMeter and mAPMeter

### PR DESCRIPTION
**Pull request by amdegroot**
Just adding two PyTorch meters (and corresponding tests) converted from lua [torchnet](https://github.com/torchnet/torchnet):
* APMeter
* mAPMeter

### CODE REVIEW

File: torchnet/meter/apmeter.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+    `target`, and optionally a `Nx1` Tensor weight where (1) the `output`
+    contains model output scores for `N` examples and `K` classes that ought to
+    be higher when the model is more convinced that the example should be
+    positively labeled, and smaller when the model believes the example should
+    be negatively labeled (for instance, the output of a sigmoid function); (2)
+    the `target` contains only values 0 (for negative examples) and 1
+    (for positive examples); and (3) the `weight` ( > 0) represents weight for
+    each sample.
+    """
+    def __init__(self):
```
Comment (szagoruyko) [inline torchnet/meter/apmeter.py]: can you add `super(APMeter, self).__init__()` here?

File: torchnet/meter/apmeter.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+            output = torch.from_numpy(output)
+        if not torch.is_tensor(target):
+            target = torch.from_numpy(target)
+
+        if weight is not None:
+            if not torch.is_tensor(weight):
+                weight = torch.from_numpy(weight)
+            weight = weight.squeeze()
+        if output.dim() == 1:
+            output = output.view(output.size(0), 1)
```
Comment (szagoruyko) [inline torchnet/meter/apmeter.py]: either `output.view(-1, 1)` or `output.unsqueeze(-1)` would be more readable here

File: torchnet/meter/apmeter.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+        else:
+            assert output.dim() == 2, \
+                'wrong output size (should be 1D or 2D with one column \
+                per class)'
+        if target.dim() == 1:
+            target = target.view(target.size(0), 1)
```
Comment (szagoruyko) [inline torchnet/meter/apmeter.py]: same as above

File: torchnet/meter/apmeter.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+                'Weight dimension 1 should be the same as that of target'
+            assert torch.min(weight) >= 0, 'Weight should be non-negative only'
+        assert torch.equal(target**2, target), \
+            'targets should be binary (0 or 1)'
+        if self.scores.numel() > 0:
+            assert target.size(1) == self.targets.size(1), \
+                'dimensions for output should match previously added examples.'
+
+        # make sure storage is of sufficient size
+        if self.scores.storage().size() < self.scores.numel() + output.numel():
```
Comment (szagoruyko) [inline torchnet/meter/apmeter.py]: why do you use storage here? can't you work with actual tensor?

File: torchnet/meter/apmeter.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        """Returns the model's average precision for each class
+
+        Return:
+            ap (FloatTensor): 1xK tensor, with avg precision for each class k
+        """
+
+        if self.scores.numel() == 0:
+            return 0
+        ap = torch.zeros(self.scores.size(1))
+        rg = torch.range(1, self.scores.size(0)).float()
```
Comment (szagoruyko) [inline torchnet/meter/apmeter.py]: range is getting deprecated - better to use arange

File: torchnet/meter/mapmeter.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    `target`, and optionally a `Nx1` Tensor weight where (1) the `output`
+    contains model output scores for `N` examples and `K` classes that ought to
+    be higher when the model is more convinced that the example should be
+    positively labeled, and smaller when the model believes the example should
+    be negatively labeled (for instance, the output of a sigmoid function); (2)
+    the `target` contains only values 0 (for negative examples) and 1
+    (for positive examples); and (3) the `weight` ( > 0) represents weight for
+    each sample.
+    """
+    def __init__(self):
```
Comment (szagoruyko) [inline torchnet/meter/mapmeter.py]: can you add super here too?

**Comment by szagoruyko** [comment]
thanks, looks good! just a few minor comments

File: torchnet/meter/apmeter.py (hunk 4 above)
Comment (amdegroot) [inline torchnet/meter/apmeter.py]: That was mostly just to keep it consistent with the lua version.  The equivalent on the actual tensor would be `self.scores.numel()`, so in that case I could probably change the line to:  but we resize the storage of the tensor to contain

Comment (amdegroot) [inline torchnet/meter/apmeter.py]: That was mostly just to keep it consistent with the lua version.  The equivalent on the actual tensor would be `self.scores.numel()`, but when resizing the storage of the tensor to account for new input, I add `1.5*current size` space to storage (I'm assuming this is done under the assumption that `add()` will be called several times) with the hopes of reducing the number of resizes that need to occur.  I could be overlooking something though.  Let me know what you think.

**Comment by amdegroot** [comment]
Awesome, thanks for the feedback.  I am a little unsure on the issue with storage, as explained, and as far as I can tell, `torch.arange()` isn't available yet in the newest version, even though it says it is favored over `torch.range()` in the docs.

**Comment by szagoruyko** [comment]
indeed arange is not available yet. Thanks!

**Pull request merged**